File: modules/springmockk-boot3/src/main/java/com/ninjasquad/springmockk/MockkBean.java
package com.ninjasquad.springmockk;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Repeatable;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import org.springframework.context.ApplicationContext;
import org.springframework.core.annotation.AliasFor;
import org.springframework.test.context.junit4.SpringRunner;

/**
 * Annotation that can be used to add MockK mocks to a Spring {@link ApplicationContext}. Can be
 * used as a class level annotation or on fields in either {@code @Configuration} classes,
 * or test classes that are run with the {@link SpringRunner}.
 * <p>
 * Mocks can be registered by type or by {@link #name() bean name}. When registered by
 * type, any existing single bean of a matching type (including subclasses) in the context
 * will be replaced by the mock. When registered by name, an existing bean can be
 * specifically targeted for replacement by a mock. In either case, if no existing bean is
 * defined a new one will be added. Dependencies that are known to the application context
 * but are not beans (such as those
 * {@link org.springframework.beans.factory.config.ConfigurableListableBeanFactory#registerResolvableDependency(Class, Object)
 * registered directly}) will not be found and a mocked bean will be added to the context
 * alongside the existing dependency.
 * <p>
 * When {@code @MockkBean} is used on a field, as well as being registered in the
 * application context, the mock will also be injected into the field. Typical usage might
 * be: <pre class="code">
 * &#064;RunWith(SpringRunner.class)
 * class ExampleTests {
 *
 *     &#064;MockkBean
 *     private lateinit var service: ExampleService
 *
 *     &#064;Autowired
 *     private lateinit var userOfService: UserOfService
 *
 *     &#064;Test
 *     void testUserOfService() {
 *         every { service.greet() } returns "Hello"
 *         val actual = userOfService.makeUse()
 *         assertThat(actual).isEqualTo("Was: Hello")
 *     }
 *
 *     &#064;Configuration
 *     &#064;Import(UserOfService::class) // A &#064;Component injected with ExampleService
 *     class Config {
 *     }
 *
 *
 * }
 * </pre> If there is more than one bean of the requested type, qualifier metadata must be
 * specified at field level: <pre class="code">
 * &#064;RunWith(SpringRunner.class)
 * class ExampleTests {
 *
 *     &#064;MockkBean
 *     &#064;Qualifier("example")
 *     private lateinit var service: ExampleService
 *
 *     ...
 * }
 * </pre>
 * <p>
 * This annotation is {@code @Repeatable} and may be specified multiple times when working
 * with Java 8 or contained within an {@link MockkBeans @MockkBeans} annotation.
 *
 * @author Phillip Webb
 * @author JB Nizet
 * @see MockkPostProcessor
 */
@Target({ElementType.TYPE, ElementType.FIELD })
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Repeatable(MockkBeans.class)
public @interface MockkBean {

    /**
     * The name of the bean to register or replace. If not specified the name will either
     * be generated or, if the mock replaces an existing bean, the existing name will be
     * used.
     * @return the name of the bean
     */
    String name() default "";

    /**
     * The classes to mock. This is an alias of {@link #classes()} which can be used for
     * brevity if no other attributes are defined. See {@link #classes()} for details.
     * @return the classes to mock
     */
    @AliasFor("classes")
    Class<?>[] value() default {};

    /**
     * The classes to mock. Each class specified here will result in a mock being created
     * and registered with the application context. Classes can be omitted when the
     * annotation is used on a field.
     * <p>
     * When {@code @MockkBean} also defines a {@code name} this attribute can only contain
     * a single value.
     * <p>
     * If this is the only specified attribute consider using the {@code value} alias
     * instead.
     * @return the classes to mock
     */
    @AliasFor("value")
    Class<?>[] classes() default {};

    /**
     * Any extra interfaces that should also be declared on the mock.
     * @return any extra interfaces
     */
    Class<?>[] extraInterfaces() default {};

    /**
     * The clear mode to apply to the mock bean. The default is {@link MockkClear#AFTER}
     * meaning that mocks are automatically reset after each test method is invoked.
     * @return the clear mode
     */
    MockkClear clear() default MockkClear.AFTER;

    /**
     * Specifies if the created mock will be relaxed or not
     * @return true if relaxed, false otherwise
     */
    boolean relaxed() default false;

    /**
     * Specifies if the created mock will have relaxed <code>Unit</code>-returning functions
     * @return true if relaxed, false otherwise
     */
    boolean relaxUnitFun() default false;
}

File: modules/springmockk/src/main/kotlin/com/ninjasquad/springmockk/MockkBean.kt
package com.ninjasquad.springmockk

import org.springframework.core.annotation.AliasFor
import org.springframework.test.context.bean.override.BeanOverride
import kotlin.reflect.KClass

/**
 * `@MockkBean` is an annotation that can be used in test classes to
 * override a bean in the test's [ApplicationContext][org.springframework.context.ApplicationContext]
 * with a MockK mock.
 *
 * `@MockkBean` can be applied in the following ways.
 *  - On a non-static field in a test class or any of its superclasses.
 *  - On a non-static field in an enclosing class for a `@Nested` test class
 *    or in any class in the type hierarchy or enclosing class hierarchy above the
 *    `@Nested` test class.
 *  - At the type level on a test class or any superclass or implemented interface
 *    in the type hierarchy above the test class.
 *  - At the type level on an enclosing class for a `@Nested` test class
 *    or on any class or interface in the type hierarchy or enclosing class hierarchy
 *    above the `@Nested` test class.
 *
 * When `@MockkBean` is declared on a field, the bean to mock is inferred
 * from the type of the annotated field. If multiple candidates exist in the
 * `ApplicationContext`, a `@field:Qualifier` annotation can be declared
 * on the field to help disambiguate. In the absence of a `@Qualifier`
 * annotation, the name of the annotated field will be used as a *fallback
 * qualifier*. Alternatively, you can explicitly specify a bean name to mock
 * by setting the [value] or [name] property.
 *
 * When `@MockkBean` is declared at the type level, the type of bean
 * (or beans) to mock must be supplied via the [types] property.
 * If multiple candidates exist in the `ApplicationContext`, you can
 * explicitly specify a bean name to mock by setting the [name]
 * property. Note, however, that the [types] attribute must contain a
 * single type if an explicit bean `name` is configured.
 *
 * A bean will be created if a corresponding bean does not exist. However, if
 * you would like for the test to fail when a corresponding bean does not exist,
 * you can set the [enforceOverride] attribute to `true`
 * &mdash; for example, `@MockkBean(enforceOverride = true)`.
 *
 * Dependencies that are known to the application context but are not beans
 * (such as those
 * [registered directly][org.springframework.beans.factory.config.ConfigurableListableBeanFactory.registerResolvableDependency]) will not be found, and a mocked bean will be added to
 * the context alongside the existing dependency.
 *
 * **WARNING**: Using `@MockkBean` in conjunction with
 * `@ContextHierarchy` can lead to undesirable results since each
 * `@MockkBean` will be applied to all context hierarchy levels by default.
 * To ensure that a particular `@MockkBean` is applied to a single context
 * hierarchy level, set the [contextName] to match a
 * configured `@ContextConfiguration`
 * [name][org.springframework.test.context.ContextConfiguration.name].
 * See the Javadoc for [`@ContextHierarchy`][org.springframework.test.context.ContextHierarchy]
 * for further details and examples.
 *
 * **NOTE**: When mocking a non-singleton bean, the non-singleton
 * bean will be replaced with a singleton mock, and the corresponding bean definition
 * will be converted to a singleton. Consequently, if you mock a prototype or scoped
 * bean, the mock will be treated as a singleton. Similarly, when mocking a bean
 * created by a [FactoryBean][org.springframework.beans.factory.FactoryBean],
 * the `FactoryBean` will be replaced with a singleton mock of the type of
 * object created by the `FactoryBean`.
 *
 * There are no restrictions on the visibility of a `@MockkBean` field.
 * Such fields can therefore be `public`, `protected`, `internal`, or
 * `private` depending on the needs or coding practices of the project.
 *
 * `@MockkBean` fields and type-level `@MockkBean` declarations
 * will be inherited from an enclosing test class by default. See
 * [NestedConfiguration][org.springframework.test.context.NestedTestConfiguration]
 * for details.
 *
 * `@MockkBean` may be used as a *meta-annotation* to create custom
 * *composed annotations* &mdash; for example, to define common mock
 * configuration in a single annotation that can be reused across a test suite.
 * `@MockkBean` can also be used as a *repeatable*
 * annotation at the type level &mdash; for example, to mock several beans by
 * [name].
 *
 * @author Simon Baslé
 * @author Sam Brannen
 * @author Jean-Baptiste Nizet
 *
 * @see MockkSpyBean
 * @see org.springframework.test.context.bean.override.convention.TestBean
 */
@Target(AnnotationTarget.FIELD, AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
@MustBeDocumented
@Repeatable
@BeanOverride(MockkBeanOverrideProcessor::class)
annotation class MockkBean(

    /**
     * Alias for [name].
     *
     * Intended to be used when no other attributes are needed &mdash; for
     * example, `@MockkBean("customBeanName")`.
     *
     * @see name
     */
    @get:AliasFor("name")
    val value: String = "",

    /**
     * Name of the bean to mock.
     *
     * If left unspecified, the bean to mock is selected according to the
     * configured [types] or the annotated field's type, taking
     * qualifiers into account if necessary. See the [class-level documentation][MockkBean]
     * for details.
     * @see value
     */
    @get:AliasFor("value")
    val name: String = "",

    /**
     * One or more types to mock.
     *
     * Defaults to none.
     *
     * Each type specified will result in a mock being created and registered
     * with the `ApplicationContext`.
     *
     * Types must be omitted when the annotation is used on a field.
     *
     * When `@MockkBean` also defines a [name], this attribute
     * can only contain a single value.
     * @return the types to mock
     */
    val types: Array<KClass<*>> = [],

    /**
     * The name of the context hierarchy level in which this `@MockkBean`
     * should be applied.
     *
     * Defaults to an empty string which indicates that this `@MockkBean`
     * should be applied to all application contexts.
     *
     * If a context name is configured, it must match a name configured via
     * `@ContextConfiguration(name=...)`.
     *
     * @see org.springframework.test.context.ContextHierarchy
     * @see org.springframework.test.context.ContextConfiguration.name
     */
    val contextName: String = "",

    /**
     * Extra interfaces that should also be declared by the mock.
     *
     * Defaults to none.
     *
     * @return any extra interfaces
     */
    val extraInterfaces: Array<KClass<*>> = [],

    /**
     * The clear mode to apply to the mock.
     *
     * The default is [MockkClear.AFTER] meaning that mocks are
     * automatically reset after each test method is invoked.
     *
     * @return the clear mode
     */
    val clear: MockkClear = MockkClear.AFTER,

    /**
     * Whether to require the existence of the bean being mocked.
     *
     * Defaults to `false` which means that a mock will be created if a
     * corresponding bean does not exist.
     *
     * Set to `true` to cause an exception to be thrown if a corresponding
     * bean does not exist.
     *
     * @see org.springframework.test.context.bean.override.BeanOverrideStrategy.REPLACE_OR_CREATE
     * @see org.springframework.test.context.bean.override.BeanOverrideStrategy.REPLACE
     */
    val enforceOverride: Boolean = false,

    /**
     * Specifies if the created mock will be relaxed or not
     * @return true if relaxed, false otherwise
     *
     * @see https://mockk.io/#relaxed-mock
     */
    val relaxed: Boolean = false,

    /**
     * Specifies if the created mock will have relaxed `Unit`-returning functions
     * @return true if relaxed, false otherwise
     *
     * @see https://mockk.io/#mock-relaxed-for-functions-returning-unit
     */
    val relaxUnitFun: Boolean = false
)
